File: GhostChat-iOS/GhostChat/Models/ControlMessage.swift
import Foundation

/// Управляющие сообщения через E2E DataChannel
/// Все типы из app.js handleControlMessage()
enum ControlMessage {
    case renegotiate(sdp: [String: Any])
    case callRequest
    case callResponse(accepted: Bool)
    case callEnd
    case callSecurityAlert(alert: [String: Any])
    case securityAlert(alert: String)
    case messageAck(counter: Int)
    case messageRead(counter: Int)
    case ready
    case pushToken(token: String)
    case notifyToken(token: String)
    case typing(isTyping: Bool)
    case capabilities(features: [String])
    case fileStart(fileId: String, name: String, size: Int64, mimeType: String, totalChunks: Int)
    case fileChunk(fileId: String, index: Int, data: String)
    case fileComplete(fileId: String)
    case roomRotate(roomId: String)
    case messageDelete(messageId: String)
    case messageEdit(messageId: String, newText: String)
    case fileRetransmit(fileId: String, indices: [Int])

    /// Создание из JSON (парсинг входящих)
    static func from(_ json: [String: Any]) -> ControlMessage? {
        guard let type = json["type"] as? String else { return nil }

        switch type {
        case "renegotiate":
            guard let sdp = json["sdp"] as? [String: Any] else { return nil }
            return .renegotiate(sdp: sdp)
        case "call-request":
            return .callRequest
        case "call-response":
            guard let accepted = json["accepted"] as? Bool else { return nil }
            return .callResponse(accepted: accepted)
        case "call-end":
            return .callEnd
        case "call-security-alert":
            guard let alert = json["alert"] as? [String: Any] else { return nil }
            return .callSecurityAlert(alert: alert)
        case "security-alert":
            guard let alert = json["alert"] as? String else { return nil }
            return .securityAlert(alert: alert)
        case "message-ack":
            guard let counter = json["c"] as? Int else { return nil }
            return .messageAck(counter: counter)
        case "message-read":
            guard let counter = json["c"] as? Int else { return nil }
            return .messageRead(counter: counter)
        case "ready":
            return .ready
        case "push-token":
            guard let token = json["token"] as? String else { return nil }
            return .pushToken(token: token)
        case "notify-token":
            guard let token = json["token"] as? String else { return nil }
            return .notifyToken(token: token)
        case "typing":
            guard let isTyping = json["isTyping"] as? Bool else { return nil }
            return .typing(isTyping: isTyping)
        case "capabilities":
            guard let features = json["features"] as? [String] else { return nil }
            return .capabilities(features: features)
        case "file-start":
            guard let fileId = json["fileId"] as? String,
                  let name = json["name"] as? String,
                  let size = json["size"] as? Int64,
                  let mimeType = json["mimeType"] as? String,
                  let totalChunks = json["totalChunks"] as? Int else { return nil }
            return .fileStart(fileId: fileId, name: name, size: size, mimeType: mimeType, totalChunks: totalChunks)
        case "file-chunk":
            guard let fileId = json["fileId"] as? String,
                  let index = json["index"] as? Int,
                  let data = json["data"] as? String else { return nil }
            return .fileChunk(fileId: fileId, index: index, data: data)
        case "file-complete":
            guard let fileId = json["fileId"] as? String else { return nil }
            return .fileComplete(fileId: fileId)
        case "room-rotate":
            guard let roomId = json["roomId"] as? String, !roomId.isEmpty else { return nil }
            return .roomRotate(roomId: roomId)
        case "message-delete":
            guard let messageId = json["messageId"] as? String, !messageId.isEmpty else { return nil }
            return .messageDelete(messageId: messageId)
        case "message-edit":
            guard let messageId = json["messageId"] as? String, !messageId.isEmpty,
                  let newText = json["newText"] as? String else { return nil }
            return .messageEdit(messageId: messageId, newText: newText)
        case "file-retransmit":
            guard let fileId = json["fileId"] as? String,
                  let indices = json["indices"] as? [Int] else { return nil }
            return .fileRetransmit(fileId: fileId, indices: indices)
        default:
            return nil
        }
    }

    /// Сериализация для отправки
    func toJSON() -> [String: Any] {
        var json: [String: Any]
        switch self {
        case .renegotiate(let sdp):
            json = ["type": "renegotiate", "sdp": sdp]
        case .callRequest:
            json = ["type": "call-request"]
        case .callResponse(let accepted):
            json = ["type": "call-response", "accepted": accepted]
        case .callEnd:
            json = ["type": "call-end"]
        case .callSecurityAlert(let alert):
            json = ["type": "call-security-alert", "alert": alert]
        case .securityAlert(let alert):
            json = ["type": "security-alert", "alert": alert]
        case .messageAck(let counter):
            json = ["type": "message-ack", "c": counter]
        case .messageRead(let counter):
            json = ["type": "message-read", "c": counter]
        case .ready:
            json = ["type": "ready"]
        case .pushToken(let token):
            json = ["type": "push-token", "token": token]
        case .notifyToken(let token):
            json = ["type": "notify-token", "token": token]
        case .typing(let isTyping):
            json = ["type": "typing", "isTyping": isTyping]
        case .capabilities(let features):
            json = ["type": "capabilities", "features": features]
        case .fileStart(let fileId, let name, let size, let mimeType, let totalChunks):
            json = ["type": "file-start", "fileId": fileId, "name": name, "size": size, "mimeType": mimeType, "totalChunks": totalChunks]
        case .fileChunk(let fileId, let index, let data):
            json = ["type": "file-chunk", "fileId": fileId, "index": index, "data": data]
        case .fileComplete(let fileId):
            json = ["type": "file-complete", "fileId": fileId]
        case .roomRotate(let roomId):
            json = ["type": "room-rotate", "roomId": roomId]
        case .messageDelete(let messageId):
            json = ["type": "message-delete", "messageId": messageId]
        case .messageEdit(let messageId, let newText):
            json = ["type": "message-edit", "messageId": messageId, "newText": newText]
        case .fileRetransmit(let fileId, let indices):
            json = ["type": "file-retransmit", "fileId": fileId, "indices": indices]
        }
        // Web клиент проверяет _ctrl чтобы отличить control от текста
        json["_ctrl"] = true
        return json
    }
}

// MARK: - Data Hex Utilities

extension Data {
    /// Инициализация из hex-строки (для push token)
    init?(hexString: String) {
        let len = hexString.count / 2
        var data = Data(capacity: len)
        var index = hexString.startIndex
        for _ in 0..<len {
            let nextIndex = hexString.index(index, offsetBy: 2)
            guard let byte = UInt8(hexString[index..<nextIndex], radix: 16) else { return nil }
            data.append(byte)
            index = nextIndex
        }
        self = data
    }

    /// Преобразование в hex-строку
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }
}

File: GhostChat-Android/app/src/main/java/com/ghost/chat/models/ControlMessage.kt
package com.ghost.chat.models

import org.json.JSONArray
import org.json.JSONObject

/// Control messages through E2E DataChannel
/// All types from app.js handleControlMessage()
sealed class ControlMessage {
    data class Renegotiate(val sdp: JSONObject) : ControlMessage()
    data object CallRequest : ControlMessage()
    data class CallResponse(val accepted: Boolean) : ControlMessage()
    data object CallEnd : ControlMessage()
    data class CallSecurityAlert(val alert: JSONObject) : ControlMessage()
    data class SecurityAlert(val alert: String) : ControlMessage()
    data class MessageAck(val counter: Int) : ControlMessage()
    data class MessageRead(val counter: Int) : ControlMessage()
    data object Ready : ControlMessage()
    data class PushToken(val token: String) : ControlMessage()
    data class NotifyToken(val token: String) : ControlMessage()
    data class Typing(val isTyping: Boolean) : ControlMessage()
    data class Capabilities(val features: List<String>) : ControlMessage()
    data class FileStart(val fileId: String, val name: String, val size: Long, val mimeType: String, val totalChunks: Int) : ControlMessage()
    data class FileChunk(val fileId: String, val index: Int, val data: String) : ControlMessage()
    data class FileComplete(val fileId: String) : ControlMessage()
    data class RoomRotate(val roomId: String) : ControlMessage()
    data class FileRetransmit(val fileId: String, val indices: List<Int>) : ControlMessage()
    data class MessageDelete(val messageId: String) : ControlMessage()
    data class MessageEdit(val messageId: String, val newText: String) : ControlMessage()

    /** Parse from JSON (incoming) */
    companion object {
        fun from(json: JSONObject): ControlMessage? {
            return when (json.optString("type", "")) {
                "renegotiate" -> {
                    val sdp = json.optJSONObject("sdp") ?: return null
                    Renegotiate(sdp)
                }
                "call-request" -> CallRequest
                "call-response" -> {
                    val accepted = json.optBoolean("accepted", false)
                    CallResponse(accepted)
                }
                "call-end" -> CallEnd
                "call-security-alert" -> {
                    val alert = json.optJSONObject("alert") ?: return null
                    CallSecurityAlert(alert)
                }
                "security-alert" -> {
                    val alert = json.optString("alert", "")
                    if (alert.isEmpty()) return null
                    SecurityAlert(alert)
                }
                "message-ack" -> {
                    val counter = json.optInt("c", -1)
                    if (counter < 0) return null
                    MessageAck(counter)
                }
                "message-read" -> {
                    val counter = json.optInt("c", -1)
                    if (counter < 0) return null
                    MessageRead(counter)
                }
                "ready" -> Ready
                "push-token" -> {
                    val token = json.optString("token", "")
                    if (token.isEmpty()) return null
                    PushToken(token)
                }
                "notify-token" -> {
                    val token = json.optString("token", "")
                    if (token.isEmpty()) return null
                    NotifyToken(token)
                }
                "typing" -> {
                    val isTyping = json.optBoolean("isTyping", false)
                    Typing(isTyping)
                }
                "capabilities" -> {
                    val arr = json.optJSONArray("features") ?: return null
                    val features = (0 until arr.length()).map { arr.getString(it) }
                    Capabilities(features)
                }
                "file-start" -> {
                    val fileId = json.optString("fileId", "")
                    val name = json.optString("name", "")
                    val size = json.optLong("size", -1)
                    val mimeType = json.optString("mimeType", "")
                    val totalChunks = json.optInt("totalChunks", -1)
                    if (fileId.isEmpty() || name.isEmpty() || size < 0 || totalChunks < 0) return null
                    FileStart(fileId, name, size, mimeType, totalChunks)
                }
                "file-chunk" -> {
                    val fileId = json.optString("fileId", "")
                    val index = json.optInt("index", -1)
                    val data = json.optString("data", "")
                    if (fileId.isEmpty() || index < 0 || data.isEmpty()) return null
                    FileChunk(fileId, index, data)
                }
                "file-complete" -> {
                    val fileId = json.optString("fileId", "")
                    if (fileId.isEmpty()) return null
                    FileComplete(fileId)
                }
                "room-rotate" -> {
                    val roomId = json.optString("roomId", "")
                    if (roomId.isEmpty()) return null
                    RoomRotate(roomId)
                }
                "file-retransmit" -> {
                    val fileId = json.optString("fileId", "")
                    val indicesArr = json.optJSONArray("indices") ?: return null
                    if (fileId.isEmpty()) return null
                    val indices = (0 until indicesArr.length()).map { indicesArr.getInt(it) }
                    FileRetransmit(fileId, indices)
                }
                "message-delete" -> {
                    val messageId = json.optString("messageId", "")
                    if (messageId.isEmpty()) return null
                    MessageDelete(messageId)
                }
                "message-edit" -> {
                    val messageId = json.optString("messageId", "")
                    val newText = json.optString("newText", "")
                    if (messageId.isEmpty()) return null
                    MessageEdit(messageId, newText)
                }
                else -> null
            }
        }
    }

    /** Serialize for sending */
    fun toJSON(): JSONObject {
        val json = when (this) {
            is Renegotiate -> JSONObject().apply {
                put("type", "renegotiate")
                put("sdp", sdp)
            }
            is CallRequest -> JSONObject().apply { put("type", "call-request") }
            is CallResponse -> JSONObject().apply {
                put("type", "call-response")
                put("accepted", accepted)
            }
            is CallEnd -> JSONObject().apply { put("type", "call-end") }
            is CallSecurityAlert -> JSONObject().apply {
                put("type", "call-security-alert")
                put("alert", alert)
            }
            is SecurityAlert -> JSONObject().apply {
                put("type", "security-alert")
                put("alert", alert)
            }
            is MessageAck -> JSONObject().apply {
                put("type", "message-ack")
                put("c", counter)
            }
            is MessageRead -> JSONObject().apply {
                put("type", "message-read")
                put("c", counter)
            }
            is Ready -> JSONObject().apply { put("type", "ready") }
            is PushToken -> JSONObject().apply {
                put("type", "push-token")
                put("token", token)
            }
            is NotifyToken -> JSONObject().apply {
                put("type", "notify-token")
                put("token", token)
            }
            is Typing -> JSONObject().apply {
                put("type", "typing")
                put("isTyping", isTyping)
            }
            is Capabilities -> JSONObject().apply {
                put("type", "capabilities")
                put("features", JSONArray(features))
            }
            is FileStart -> JSONObject().apply {
                put("type", "file-start")
                put("fileId", fileId)
                put("name", name)
                put("size", size)
                put("mimeType", mimeType)
                put("totalChunks", totalChunks)
            }
            is FileChunk -> JSONObject().apply {
                put("type", "file-chunk")
                put("fileId", fileId)
                put("index", index)
                put("data", data)
            }
            is FileComplete -> JSONObject().apply {
                put("type", "file-complete")
                put("fileId", fileId)
            }
            is FileRetransmit -> JSONObject().apply {
                put("type", "file-retransmit")
                put("fileId", fileId)
                put("indices", JSONArray(indices))
            }
            is RoomRotate -> JSONObject().apply {
                put("type", "room-rotate")
                put("roomId", roomId)
            }
            is MessageDelete -> JSONObject().apply {
                put("type", "message-delete")
                put("messageId", messageId)
            }
            is MessageEdit -> JSONObject().apply {
                put("type", "message-edit")
                put("messageId", messageId)
                put("newText", newText)
            }
        }
        // Web клиент проверяет _ctrl чтобы отличить control от текста
        json.put("_ctrl", true)
        return json
    }
}
